[Form control refresh] Text/decoration colors are incorrect for button-like controls in certain cases

lilyspiniolas opened the pull request:
#### 03eba551b315785466c30c67be0dba1fcb3e1962
<pre>
[Form control refresh] Text/decoration colors are incorrect for button-like controls in certain cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=295170">https://bugs.webkit.org/show_bug.cgi?id=295170</a>
<a href="https://rdar.apple.com/153937240">rdar://153937240</a>

Reviewed by Abrar Rahman Protyasha.

* The text for disabled submit buttons now uses tertiary label color
  when the window is active, and matches the text color of normal
  buttons when the window is inactive. This is achieved by setting
  a new style flag `insideDisabledSubmitButton` when appropriate during
  style adjustment. This flag is checked for during text painting, and
  when present, the text color is adjusted as necessary.

* For &lt;select&gt; buttons and non-submit buttons, tint color is no longer
  used as the text color or decoration color on any platform.

* The text/decoration color of &lt;select&gt; buttons now matches the
  color of typical buttons.

* When the Form control refresh is supported, clicking a button or otherwise
  using CSS value &quot;ActiveButtonText&quot; no longer creates a `NSButtonCell` in
  order to determine the text color to use. Instead, the typical button text
  color is used.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/form-control-refresh/button-like-controls-text-color-matches-expected.txt: Added.
* LayoutTests/fast/forms/form-control-refresh/button-like-controls-text-color-matches.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-appearance-matches-button-window-inactive-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-appearance-matches-button-window-inactive.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-text-color-disabled-state-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-text-color-disabled-state.html: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::disabledSubmitButtonTextColor const):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustMenuListStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::buttonTextColor const):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustMenuListButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
(WebCore::RenderThemeCocoa::disabledSubmitButtonTextColor const):
(WebCore::adjustButtonLikeControlStyleForVectorBasedControls): Deleted.
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresRepaint):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideDisabledSubmitButton const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideDisabledSubmitButton):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/296886@main">https://commits.webkit.org/296886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04ab8f3650e641377ffdb5d51131fe739898cb4f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109784 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115805 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60019 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111747 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83425 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112732 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24003 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63888 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23382 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17012 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59599 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118597 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36823 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27278 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92433 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37196 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92256 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23522 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37222 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14967 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32660 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36718 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42188 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36378 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39720 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38087 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/rendering/TextPaintStyle.cpp:
```diff
@@ -101,6 +101,17 @@ TextPaintStyle computeTextPaintStyle(const RenderText& renderer, const RenderSty
+        if (page && page->focusController().isActive()) {
+            OptionSet<StyleColorOptions> options;
+            if (page->settings().useSystemAppearance())
+                options.add(StyleColorOptions::UseSystemAppearance);
```
This isn't being used anywhere it seems?

aprotyas commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -2186,21 +2159,32 @@ static void adjustButtonLikeControlStyleForVectorBasedControls(RenderStyle& styl
+
+    const auto styleColorOptions = element->document().styleColorOptions(&style);
+
+    auto adjustStyleForSubmitButton = [&]() {
```
You can drop the empty parameter list in C++23-onwards
```suggestion
    auto adjustStyleForSubmitButton = [&] {
```

aprotyas commented on Source/WebCore/rendering/RenderTheme.h:
```diff
@@ -261,6 +261,8 @@ class RenderTheme {
     virtual void adjustTextControlInnerPlaceholderStyle(RenderStyle&, const RenderStyle&, const Element*) const { }
     virtual void adjustTextControlInnerTextStyle(RenderStyle&, const RenderStyle&, const Element*) const { }
 
+    virtual Color disabledSubmitButtonTextColor() const { return Color::black; }
```
This can be a `constexpr` method.

aprotyas replied:
(optional nit, of course)

lilyspiniolas committed:
[Form control refresh] Text/decoration colors are incorrect for button-like controls in certain cases
https://bugs.webkit.org/show_bug.cgi?id=295170
rdar://153937240

Reviewed by Abrar Rahman Protyasha.

* The text for disabled submit buttons now uses tertiary label color
  when the window is active, and matches the text color of normal
  buttons when the window is inactive. This is achieved by setting
  a new style flag `insideDisabledSubmitButton` when appropriate during
  style adjustment. This flag is checked for during text painting, and
  when present, the text color is adjusted as necessary.

* For <select> buttons and non-submit buttons, tint color is no longer
  used as the text color or decoration color on any platform.

* The text/decoration color of <select> buttons now matches the
  color of typical buttons.

* When the Form control refresh is supported, clicking a button or otherwise
  using CSS value "ActiveButtonText" no longer creates a `NSButtonCell` in
  order to determine the text color to use. Instead, the typical button text
  color is used.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/form-control-refresh/button-like-controls-text-color-matches-expected.txt: Added.
* LayoutTests/fast/forms/form-control-refresh/button-like-controls-text-color-matches.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-appearance-matches-button-window-inactive-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-appearance-matches-button-window-inactive.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-text-color-disabled-state-expected.html: Added.
* LayoutTests/fast/forms/form-control-refresh/submit-button-text-color-disabled-state.html: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::disabledSubmitButtonTextColor const):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::adjustMenuListStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::buttonTextColor const):
(WebCore::RenderThemeCocoa::adjustButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::adjustMenuListButtonStyleForVectorBasedControls const):
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
(WebCore::RenderThemeCocoa::disabledSubmitButtonTextColor const):
(WebCore::adjustButtonLikeControlStyleForVectorBasedControls): Deleted.
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::RenderThemeMac::systemColor const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresRepaint):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideDisabledSubmitButton const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideDisabledSubmitButton):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: https://commits.webkit.org/296886@main
